tests/acceptance/common.sh

```diff
@@ -0,0 +1,138 @@
+function log() {
+    local level="${2:-INFO}"
+    echo "[$(TZ=UTC date --rfc-3339=ns) ${level} ${BATS_TEST_NAME}] $1"
+}
+
+function logError() {
+    log $1 'ERROR'
+}
+
+function logWarn() {
+    log $1 'WARN'
+}
+
+function testAzureSecret() {
+    local role_name="$1"
+    local subscription_id="$2"
+    local resource_group_name="$3"
+    local vault_role_name="$4"
+    local config_dir="$5"
+    local engine_name="$6"
+    local ttl=5
+
+    log "Creating Azure secret, azure_role='${role_name}', vault_role='${vault_role_name}'"
+    vault write "${engine_name}/roles/${vault_role_name}" ttl="${ttl}" azure_roles=-<<EOF
+[
+    {
+        "role_name": "${role_name}",
+        "scope":  "/subscriptions/${subscription_id}/resourceGroups/${resource_group_name}"
+    }
+]
+EOF
+
+    local secret_file="${config_dir}/secret-${vault_role_name}.json"
+    vault read "${engine_name}/creds/${vault_role_name}" -format=json > ${secret_file}
+
+    sp_id=$(cat "${secret_file}" | jq -er .data.client_id)
+    log "Secret created successfully, azure_role='${role_name}', vault_role='${vault_role_name}', file=${secret_file}"
+
+    if !assertSPExistence ; then
+        exit 1
+    fi
+
+    if !waitLeaseExpiration ${role_name} ${vault_role_name} ${sp_id} ${ttl}; then
+        exit 1
+    fi
+
+    log "Test completed successfully, azure_role='${role_name}', vault_role='${vault_role_name}'"
+}
+
+function waitLeaseExpiration() {
+    local role_name="${1}"
+    local vault_role_name="${2}"
+    local sp_id="${3}"
+    local ttl="${4}"
+    log "Waiting for lease expiration, azure_role='${role_name}', vault_role='${vault_role_name}'"
+    sleep ${ttl}
+    local tries=0
+    # wait for the service principal to expire and be removed by Vault - adds a 60 second buffer to the ttl.
+    until ! az ad sp show --id "${sp_id}" &> /dev/null
+    do
+        if [[ "${tries}" -ge 60 ]]; then
+            logError "Vault failed to remove service principal ${sp_id}, ttl=${ttl}"
+            return 1
+        fi
+        ((++tries))
+        sleep 1
+    done
+}
+
+function assertSPExistence() {
+    local sp_id="${1}"
+    local found=''
+    for n in {0..30}
+    do
+         if ! az ad sp show --id "${sp_id}" >&1 /dev/null ; then
+            logWarn "Failed to check service principal exists for ID ${sp_id}"
+            sleep 1
+            continue
+         fi
```

Comment:
Doesn't this need another redirect to send stdout to /dev/null?

```suggestion
         if ! az ad sp show --id "${sp_id}" >&1 > /dev/null ; then
            logWarn "Failed to check service principal exists for ID ${sp_id}"
            sleep 1
            continue
         fi
```

My earlier suggestion around using bats' `run` could look something like this too, since run always returns 0:

```suggestion
         run az ad sp show --id "${sp_id}"
         if [ $status -ne 0 ] ; then
            logWarn "Failed to check service principal exists for ID ${sp_id}: $output"
            sleep 1
            continue
         fi
```

That strikes me as being a little closer to the style you'd see in a non-bash test, which is what I was thinking as far as being more readable. YMMV.

Reply:
Definitely a good point, but I feel that these functions are common to the tests and could be used outside of bats, so I think I would prefer to keep it as-is.